### PR TITLE
add 3dt config

### PR DESCRIPTION
**Pull request by darkonie**

### CODE REVIEW

File: api/config.go [hunk 1]
```diff
@@ -24,37 +28,43 @@ var (
 
 // Config structure is a main config object
 type Config struct {
-	Version                 string
-	Revision                string
-	MesosIPDiscoveryCommand string
-	DCOSVersion             string
-	SystemdUnits            []string
+	Version                                      string `json:"-"`
```
Comment (malnick) [inline api/config.go]: nit: prefer `omitempty` over `-` because we use `omitempty` in a lot of our other Mesosphere golang code.

**Comment by malnick** [comment]
Minor comment regarding https://github.com/darkonie/3dt/blob/ee947dd82a16ad112de1b7745201b6f1ac1bdc0c/api/config.go#L164

It would be cleaner to do the ip detect and other default logic first, set those to variables, then simply:

```
return Config{
    thing: value,
    foo:     bar,
    }, nil
```

Versus having a lot of:
```
config.Thing = foo
...
```

File: api/config.go [hunk 2]
```diff
@@ -147,5 +157,50 @@ func LoadDefaultConfig(args []string) (config Config, err error) {
...
 }
+
+func readConfigFile(configPath string, originalConfig Config) (Config, error) {
+	// load config
+	configContent, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return originalConfig, err
+	}
+
+	var configJSON map[string]interface{}
```
Comment (malnick) [inline api/config.go]: You can unmarshal directly to your `Config{}` struct, the JSON tags enable validation on the fly.

See: https://github.com/dcos/dcos-signal/blob/master/config/config.go#L96

You can actually do this over and over again on the same struct pointer if you have more than one config file.

File: api/config.go [hunk 3]
```diff
@@ -147,5 +157,50 @@ func LoadDefaultConfig(args []string) (config Config, err error) {
...
+	if err := json.Unmarshal(configContent, &configJSON); err != nil {
+		return originalConfig, err
+	}
+
+	if len(configJSON) == 0 {
+		return originalConfig, fmt.Errorf("Empty config file provided: %s", configPath)
+	}
+
+	// validate every key in a config file
+	for k, v := range configJSON {
```
Comment (malnick) [inline api/config.go]: See above comment regarding unmarshalling directly to `Config{}`

File: api/config.go [hunk 4]
```diff
@@ -147,5 +157,50 @@ func LoadDefaultConfig(args []string) (config Config, err error) {
...
+	if err != nil {
+		return originalConfig, err
+	}
+
+	var configJSON map[string]interface{}
+	if err := json.Unmarshal(configContent, &configJSON); err != nil {
+		return originalConfig, err
+	}
+
+	if len(configJSON) == 0 {
```
Comment (alberts) [inline api/config.go]: this seems strange... let's assume that we have a config file.

**Comment by darkonie** [comment]
@malnick addressed comments, please do the review

File: api/config.go [hunk 5]
```diff
@@ -147,5 +157,53 @@ func LoadDefaultConfig(args []string) (config Config, err error) {
...
+		if err != nil {
+			return fmt.Errorf("Error validating config field: %s. %s", k, err)
+		}
+
+		tmpConfig := Config{}
+		if err := json.Unmarshal(configItem, &tmpConfig); err != nil {
+			return fmt.Errorf("Error validating config field: %s. %s", k, err)
+		}
+
+		if reflect.DeepEqual(tmpConfig, Config{}) {
```
Comment (jdef) [inline api/config.go]: this works fine until your fields are allowed to contain zero-values. in that case `DeepEqual` will return true even though it's not an error.

please seriously consider using a 3rd party json schema/validation lib to help with this.

Comment (jdef) [inline api/config.go]: the other poor-man's approach would be to make all your fields pointers. then you can more easily tell when they're present vs zero-values. if you don't want the rest of the dependent code to deal with reading (possibly nil) pointers then have two versions of config: DTOConfig (with pointers) and Config (without pointers) .. and conversion funcs to translate between the two.

Comment (jdef) [inline api/config.go]: or, perhaps, define your model in protobuf and generate ser/deser with something like ffjson (https://github.com/pquerna/ffjson)

Comment (jdef) [inline api/config.go]: hm .. I don't see how/where ffjson actually helps your use case. So I'm back to recommending something like jsonschema (https://github.com/xeipuuv/gojsonschema) for the long term.

**Comment by darkonie** [comment]
@malnick we decided to go with gojsonschema to validate the input

**Comment by malnick** [comment]
Even better, I'll make sure to update signal to match so we have consistency.

**Comment by darkonie** [comment]
a new PR https://github.com/dcos/3dt/pull/67

**Pull request closed without merging**